agent/agent.go

```diff
@@ -362,39 +365,43 @@ func updateStructValuesHelper(pluginPtr reflect.Value, newConfig map[string]inte
 				return pluginPtr.Elem(), fmt.Errorf("unsettable field %s", configKey)
 			} else if !(pluginField.Type() == reflectedNew.Type()) {
 				return pluginPtr.Elem(), fmt.Errorf("value type mismatch for field %s", configKey)
-			} else {
-				pluginField.Set(reflectedNew)
 			}
 		}
+
+		// if no error, update all
+		for configKey, configValue := range newConfig {
+			pluginField := plugin.FieldByName(configKey) // cast new value as Value
+			reflectedNew := reflect.ValueOf(configValue)
+			pluginField.Set(reflectedNew)
+		}
 		return plugin, nil
 	}
+
 	return plugin, fmt.Errorf("could not update plugin")
 }
 
-// TODO Replace this after merge.
-func (a *Agent) AddInput(pluginName string) {
-	plugin := inputs.Inputs[pluginName]
+// StartInput adds an input plugin with default config
+func (a *Agent) StartInput(pluginName string) error {
 	inputConfig := models.InputConfig{
 		Name: pluginName,
-		// TODO rest of config?
 	}
-	runningPlugin := models.NewRunningInput(plugin(), &inputConfig)
-	a.Config.Inputs = append(a.Config.Inputs, runningPlugin)
+	err := a.RunSingleInput(&inputConfig, nil, a.Context)
```

Comment:
Normally better practice to just do ```return a.RunSingleInput(&inputConfig, nil, a.Context)``` if you're not doing anything with the variable.